### PR TITLE
Refix ShenandoahPurgeSATBTask

### PR DESCRIPTION
**Pull request by zhengyu123**
My early fix was incomplete and incorrect.

1) ShenandoahPurgeSATBTask needs to change thread claim token for iterating threads in parallel, this was missing in early change.
2) Threads::threads_do() -> Threads::possibly_parallel_threads_do() was incorrect, the later version only iterates Java and VM threads. This part of changes are reverted.
3) A minor change: declaring _trashed_oops as volatile as it is a shared variable.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Reviewers
 * @earthling-amzn (no known github.com user name / role)
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/shenandoah pull/25/head:pull/25` \
`$ git checkout pull/25`

Update a local copy of the PR: \
`$ git checkout pull/25` \
`$ git pull https://git.openjdk.java.net/shenandoah pull/25/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25`

View PR using the GUI difftool: \
`$ git pr show -t 25`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/shenandoah/pull/25.diff">https://git.openjdk.java.net/shenandoah/pull/25.diff</a>

</details>

### CODE REVIEW

**Review by earthling-amzn — APPROVED** [review]
Thanks again for setting this straight.

**Comment by zhengyu123** [comment]
> Thanks again for setting this straight.

@earthling-amzn Thanks for reviewing and sorry for the noise.

**Comment by earthling-amzn** [comment]
Can we move this PR forward?

**Comment by zhengyu123** [comment]
> Can we move this PR forward?

Roman is on PTO this week, I will ping him next week.

File: src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp [hunk 1]
```diff
@@ -92,12 +96,14 @@ class ShenandoahPurgeSATBTask : public AbstractGangTask {
   ShenandoahObjToScanQueueSet* _mark_queues;
 
 public:
-  size_t _trashed_oops;
+  volatile size_t _trashed_oops;
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp:99]: If it the intention is to make shared access to the field safe, then C++ volatile does not (fully) achieve this. You also need to use Atomic::inc() and similar to change/read it.

**Review by rkennke — APPROVED** [review]
Looks good, only a minor thing about shared access to _trashed_oops.

File: src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp (hunk 1 above)
Comment (zhengyu123) [reply src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp:99]: Adding volatile keyword, mainly just for coding convention on shared variable. It uses Atomic::add() to update variable, so it is correct.

**Pull request closed without merging**